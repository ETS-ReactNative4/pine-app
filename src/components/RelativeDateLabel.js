import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment-timezone';

import { formatTime, REGION_LOCALE } from '../localization';
import StyledText from './StyledText';

const UPDATE_INTERVAL = 60 * 1000; // Every minute.

// Use English as language until other languages are supported.
const DATE_LOCALE = `en-${REGION_LOCALE}`;

moment.updateLocale('en', {
  relativeTime: {
    future: 'in %s',
    past: '%s ago',
    s: 'Now',
    ss: 'Now',
    m: '1 minute',
    mm: '%d minutes',
    h: '1 hour',
    hh: '%d hours',
    d: '1 day',
    dd: '%d days',
    M: '1 month',
    MM: '%d months',
    y: '1 year',
    yy: '%d years'
  }
});

export default class RelativeDateLabel extends Component {
  componentDidMount() {
    // Auto-update component every minute to update the relative time label.
    if (this._shouldForceUpdate()) {
      this._interval = setInterval(this._forceUpdate.bind(this), UPDATE_INTERVAL);
    }
  }

  componentWillUnmount() {
    clearInterval(this._interval);
  }

  _forceUpdate() {
    this.forceUpdate();

    if (!this._shouldForceUpdate()) {
      clearInterval(this._interval);
    }
  }

  _shouldForceUpdate() {
    const now = moment();
    const date = moment(this.props.date);
    const differenceInMinutes = now.diff(date, 'minutes');

    // Only auto-update if date was less than 60 minutes ago.
    return Math.abs(differenceInMinutes) < 60;
  }

  _getRelativeDate(date) {
    const withoutSuffix = true;
    const relativeDate = moment(date).fromNow(withoutSuffix);

    if (relativeDate === 'Now') {
      return 'Just now';
    }

    return `${relativeDate} ago`;
  }

  _getDateString(date) {
    const momentDate = moment(date);
    const now = moment();
    const yesterday = moment().subtract(1, 'days');
    const time = formatTime(date);

    if (momentDate.isSame(now, 'day')) {
      return this._getRelativeDate(date);
    }

    if (momentDate.isSame(yesterday, 'day')) {
      return `Yesterday, ${time}`;
    }

    if (momentDate.isSame(now, 'week')) {
      const weekday = moment.weekdays(momentDate.weekday());
      return `${weekday}, ${time}`;
    }

    if (momentDate.isSame(now, 'year')) {
      return date.toLocaleDateString(DATE_LOCALE, { month: 'long', day: 'numeric' });
    }

    return date.toLocaleDateString(DATE_LOCALE, { year: 'numeric', month: 'long', day: 'numeric' });
  }

  render() {
    const dateString = this._getDateString(this.props.date);

    return (
      <StyledText style={this.props.style}>
        {dateString}
      </StyledText>
    );
  }
}

RelativeDateLabel.propTypes = {
  style: PropTypes.any,
  date: PropTypes.object
};
