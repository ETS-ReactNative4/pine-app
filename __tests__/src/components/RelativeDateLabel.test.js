import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import moment from 'moment-timezone';
import RelativeDateLabel from '../../../src/components/RelativeDateLabel';

const RealDate = Date;

const mockDate = (fakeDate) => {
  global.Date = class extends RealDate {
    constructor(specifiedDate) {
      super(...arguments);
      return new RealDate(specifiedDate || fakeDate);
    }
  };

  global.Date.now = jest.fn(() => new RealDate(fakeDate).getTime());
};

const unmockDate = () => {
  global.Date = RealDate;
};

describe('RelativeDateLabel', () => {
  beforeAll(() => {
    mockDate('29 November, 2018, 21:59 GMT+02:00');
  });

  afterAll(() => {
    unmockDate();
  });

  it('renders correctly when date is now', () => {
    const now = new Date();

    const tree = renderer.create(
      <RelativeDateLabel date={now} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when date was 2 hours ago but still the same day', () => {
    const twoHoursAgo = moment().subtract(2, 'hours');

    const tree = renderer.create(
      <RelativeDateLabel date={twoHoursAgo.toDate()} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when date was yesterday', () => {
    const yesterday = moment().subtract(1, 'days');

    const tree = renderer.create(
      <RelativeDateLabel date={yesterday.toDate()} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when date was 2 days ago but still the same week', () => {
    const twoDaysAgo = moment().subtract(2, 'days');

    const tree = renderer.create(
      <RelativeDateLabel date={twoDaysAgo.toDate()} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when date was 1 week ago', () => {
    const oneWeekAgo = moment().subtract(1, 'weeks');

    const tree = renderer.create(
      <RelativeDateLabel date={oneWeekAgo.toDate()} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when date was 2 weeks ago but still the same month', () => {
    const twoWeekAgo = moment().subtract(2, 'weeks');

    const tree = renderer.create(
      <RelativeDateLabel date={twoWeekAgo.toDate()} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when date was 1 month ago but still the same year', () => {
    const oneMonthAgo = moment().subtract(1, 'months');

    const tree = renderer.create(
      <RelativeDateLabel date={oneMonthAgo.toDate()} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when date was 1 year ago', () => {
    const oneYearAgo = moment().subtract(1, 'years');

    const tree = renderer.create(
      <RelativeDateLabel date={oneYearAgo.toDate()} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
