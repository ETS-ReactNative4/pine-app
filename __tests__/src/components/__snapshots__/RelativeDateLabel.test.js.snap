// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelativeDateLabel renders correctly when date is now 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  Just now
</Text>
`;

exports[`RelativeDateLabel renders correctly when date was 1 month ago but still the same year 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  October 29
</Text>
`;

exports[`RelativeDateLabel renders correctly when date was 1 week ago 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  November 22
</Text>
`;

exports[`RelativeDateLabel renders correctly when date was 1 year ago 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  November 29, 2017
</Text>
`;

exports[`RelativeDateLabel renders correctly when date was 2 days ago but still the same week 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  Tuesday, 8:59 PM
</Text>
`;

exports[`RelativeDateLabel renders correctly when date was 2 hours ago but still the same day 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  2 hours ago
</Text>
`;

exports[`RelativeDateLabel renders correctly when date was 2 weeks ago but still the same month 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  November 15
</Text>
`;

exports[`RelativeDateLabel renders correctly when date was yesterday 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "System",
        "fontWeight": "400",
      },
      undefined,
    ]
  }
>
  Yesterday, 8:59 PM
</Text>
`;
